internal/controllers/addon_controller.go

```diff
@@ -58,35 +60,60 @@ func (r *AddonReconciler) Reconcile(
 
 	// Phase 1.
 	// Ensure wanted namespaces
-	stopAndRetry, err := r.ensureWantedNamespaces(ctx, addon)
-	if err != nil {
-		return ctrl.Result{}, fmt.Errorf("failed to ensure wanted namespaces: %w", err)
-	}
-	if stopAndRetry {
-		return ctrl.Result{
-			RequeueAfter: defaultRetryAfterTime,
-		}, nil
+	{
+		stopAndRetry, err := r.ensureWantedNamespaces(ctx, addon)
+		if err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to ensure wanted Namespaces: %w", err)
+		}
+		if stopAndRetry {
+			return ctrl.Result{
+				RequeueAfter: defaultRetryAfterTime,
+			}, nil
+		}
 	}
 
 	// Phase 2.
 	// Ensure unwanted namespaces are removed
-	err = r.ensureDeletionOfUnwantedNamespaces(ctx, addon)
-	if err != nil {
+	if err := r.ensureDeletionOfUnwantedNamespaces(ctx, addon); err != nil {
 		return ctrl.Result{}, fmt.Errorf("failed to ensure deletion of unwanted Namespaces: %w", err)
 	}
 
 	// Phase 3.
 	// Ensure OperatorGroup
-	stop, err := r.ensureOperatorGroup(ctx, log, addon)
-	if err != nil {
-		return ctrl.Result{}, fmt.Errorf("failed to ensure OperatorGroup: %w", err)
+	{
+		stop, err := r.ensureOperatorGroup(ctx, log, addon)
+		if err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to ensure OperatorGroup: %w", err)
+		}
+		if stop {
+			return ctrl.Result{}, nil
+		}
 	}
-	if stop {
-		return ctrl.Result{}, nil
+
+	// Phase 4.
+	// Ensure CatalogSource for this Addon
+	{
+		stop, retry, err := r.ensureCatalogSource(ctx, log, addon)
+		if err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to ensure CatalogSource: %w", err)
+		}
+		if stop {
+			if retry {
```

Comment:
What now? should it stop or retry?
Can we just prioritize one over the other, instead of nesting these cases?

Reply:
What do you think of this?

```diff
diff --git a/internal/controllers/addon_controller.go b/internal/controllers/addon_controller.go
index cc77f0e..91619ff 100644
--- a/internal/controllers/addon_controller.go
+++ b/internal/controllers/addon_controller.go
@@ -93,16 +93,16 @@ func (r *AddonReconciler) Reconcile(
 	// Phase 4.
 	// Ensure CatalogSource for this Addon
 	{
-		stop, retry, err := r.ensureCatalogSource(ctx, log, addon)
+		ensureResult, err := r.ensureCatalogSource(ctx, log, addon)
 		if err != nil {
 			return ctrl.Result{}, fmt.Errorf("failed to ensure CatalogSource: %w", err)
 		}
-		if stop {
-			if retry {
-				return ctrl.Result{
-					RequeueAfter: defaultRetryAfterTime,
-				}, nil
-			}
+		switch ensureResult {
+		case EnsureCatalogSourceResultRetry:
+			return ctrl.Result{
+				RequeueAfter: defaultRetryAfterTime,
+			}, nil
+		case EnsureCatalogSourceResultStop:
 			return ctrl.Result{}, nil
 		}
 	}
diff --git a/internal/controllers/phase_ensure_catalogsource.go b/internal/controllers/phase_ensure_catalogsource.go
index 8007ad5..2d38c62 100644
--- a/internal/controllers/phase_ensure_catalogsource.go
+++ b/internal/controllers/phase_ensure_catalogsource.go
@@ -16,27 +16,36 @@ import (
 	addonsv1alpha1 "github.com/openshift/addon-operator/apis/addons/v1alpha1"
 )
 
+type EnsureCatalogSourceResult int
+
+const (
+	EnsureCatalogSourceResultNil = iota
+	EnsureCatalogSourceResultStop
+	EnsureCatalogSourceResultRetry
+)
+
 const catalogSourcePublisher = "OSD Red Hat Addons"
 
 // Ensure existence of the CatalogSource specified in the given addon resource
 // returns a bool that signals the caller to stop reconciliation and retry later
 func (r *AddonReconciler) ensureCatalogSource(
-	ctx context.Context, log logr.Logger, addon *addonsv1alpha1.Addon) (stop, retry bool, err error) {
+	ctx context.Context, log logr.Logger, addon *addonsv1alpha1.Addon) (EnsureCatalogSourceResult, error) {
 	var targetNamespace, catalogSourceImage string
 
+	// stop, retry bool
 	switch addon.Spec.Install.Type {
 	case addonsv1alpha1.OwnNamespace:
 		if addon.Spec.Install.OwnNamespace == nil ||
 			len(addon.Spec.Install.OwnNamespace.Namespace) == 0 {
 			// invalid/missing configuration
 			// TODO: Move error reporting into webhook and reduce this code to a sanity check.
-			return true, false, r.reportConfigurationError(ctx, addon, ".spec.install.ownNamespace.namespace is required when .spec.install.type = OwnNamespace")
+			return EnsureCatalogSourceResultStop, r.reportConfigurationError(ctx, addon, ".spec.install.ownNamespace.namespace is required when .spec.install.type = OwnNamespace")
 		}
 		targetNamespace = addon.Spec.Install.OwnNamespace.Namespace
 		if len(addon.Spec.Install.OwnNamespace.CatalogSourceImage) == 0 {
 			// invalid/missing configuration
 			// TODO: Move error reporting into webhook and reduce this code to a sanity check.
-			return true, false, r.reportConfigurationError(ctx, addon, ".spec.install.ownNamespacee.catalogSourceImage is required when .spec.install.type = OwnNamespace")
+			return EnsureCatalogSourceResultStop, r.reportConfigurationError(ctx, addon, ".spec.install.ownNamespacee.catalogSourceImage is required when .spec.install.type = OwnNamespace")
 		}
 		catalogSourceImage = addon.Spec.Install.OwnNamespace.CatalogSourceImage
 
@@ -45,13 +54,13 @@ func (r *AddonReconciler) ensureCatalogSource(
 			len(addon.Spec.Install.AllNamespaces.Namespace) == 0 {
 			// invalid/missing configuration
 			// TODO: Move error reporting into webhook and reduce this code to a sanity check.
-			return true, false, r.reportConfigurationError(ctx, addon, ".spec.install.allNamespaces.namespace is required when .spec.install.type = AllNamespaces")
+			return EnsureCatalogSourceResultStop, r.reportConfigurationError(ctx, addon, ".spec.install.allNamespaces.namespace is required when .spec.install.type = AllNamespaces")
 		}
 		targetNamespace = addon.Spec.Install.AllNamespaces.Namespace
 		if len(addon.Spec.Install.AllNamespaces.CatalogSourceImage) == 0 {
 			// invalid/missing configuration
 			// TODO: Move error reporting into webhook and reduce this code to a sanity check.
-			return true, false, r.reportConfigurationError(ctx, addon, ".spec.install.allNamespaces.catalogSourceImage is required when .spec.install.type = AllNamespaces")
+			return EnsureCatalogSourceResultStop, r.reportConfigurationError(ctx, addon, ".spec.install.allNamespaces.catalogSourceImage is required when .spec.install.type = AllNamespaces")
 		}
 		catalogSourceImage = addon.Spec.Install.AllNamespaces.CatalogSourceImage
 
@@ -60,7 +69,7 @@ func (r *AddonReconciler) ensureCatalogSource(
 		// This should never happen, unless the schema validation is wrong.
 		// The .install.type property is set to only allow known enum values.
 		log.Error(fmt.Errorf("invalid Addon install type: %q", addon.Spec.Install.Type), "stopping Addon reconcilation")
-		return true, false, nil
+		return EnsureCatalogSourceResultStop, nil
 	}
 
 	catalogSource := &operatorsv1alpha1.CatalogSource{
@@ -79,7 +88,7 @@ func (r *AddonReconciler) ensureCatalogSource(
 	addCommonLabels(catalogSource.Labels, addon)
 
 	if err := controllerutil.SetControllerReference(addon, catalogSource, r.Scheme); err != nil {
-		return false, false, err
+		return EnsureCatalogSourceResultNil, err
 	}
 
 	var observedCatalogSource *operatorsv1alpha1.CatalogSource
@@ -87,16 +96,16 @@ func (r *AddonReconciler) ensureCatalogSource(
 		var err error
 		observedCatalogSource, err = reconcileCatalogSource(ctx, r.Client, catalogSource)
 		if err != nil {
-			return false, false, err
+			return EnsureCatalogSourceResultNil, err
 		}
 	}
 
 	if observedCatalogSource.Status.GRPCConnectionState == nil {
 		err := r.reportCatalogSourceUnreadinessStatus(ctx, addon, observedCatalogSource, ".Status.GRPCConnectionState is nil")
 		if err != nil {
-			return false, false, err
+			return EnsureCatalogSourceResultNil, err
 		}
-		return true, true, nil
+		return EnsureCatalogSourceResultRetry, nil
 	}
 	if observedCatalogSource.Status.GRPCConnectionState.LastObservedState != "READY" {
 		err := r.reportCatalogSourceUnreadinessStatus(
@@ -109,12 +118,12 @@ func (r *AddonReconciler) ensureCatalogSource(
 			),
 		)
 		if err != nil {
-			return false, false, err
+			return EnsureCatalogSourceResultNil, err
 		}
-		return true, true, err
+		return EnsureCatalogSourceResultRetry, err
 	}
 
-	return false, false, nil
+	return EnsureCatalogSourceResultNil, nil
 }
 
 // Marks Addon as unavailable because the CatalogSource is unready
diff --git a/internal/controllers/phase_ensure_catalogsource_test.go b/internal/controllers/phase_ensure_catalogsource_test.go
index 8d7b04b..de1c6a6 100644
--- a/internal/controllers/phase_ensure_catalogsource_test.go
+++ b/internal/controllers/phase_ensure_catalogsource_test.go
@@ -133,10 +133,9 @@ func TestEnsureCatalogSource_Create(t *testing.T) {
 	log := testutil.NewLogger(t)
 
 	ctx := context.Background()
-	stop, retry, err := r.ensureCatalogSource(ctx, log, addon)
+	ensureResult, err := r.ensureCatalogSource(ctx, log, addon)
 	assert.NoError(t, err)
-	assert.False(t, stop)
-	assert.False(t, retry)
+	assert.Equal(t, EnsureCatalogSourceResultNil, ensureResult)
 	c.AssertExpectations(t)
 }
 
@@ -169,10 +168,9 @@ func TestEnsureCatalogSource_Update(t *testing.T) {
 	log := testutil.NewLogger(t)
 
 	ctx := context.Background()
-	stop, retry, err := r.ensureCatalogSource(ctx, log, addon)
+	ensureResult, err := r.ensureCatalogSource(ctx, log, addon)
 	assert.NoError(t, err)
-	assert.False(t, stop)
-	assert.False(t, retry)
+	assert.Equal(t, EnsureCatalogSourceResultNil, ensureResult)
 	c.AssertExpectations(t)
 	c.AssertNumberOfCalls(t, "Get", 1)
 	c.AssertNumberOfCalls(t, "Update", 1)
```

Reply:
:+1: 
But please make the type `EnsureCatalogSourceResult` and it's const values private.

Reply:
Done! :)